README.md
### LevelDOWN for Mobile  
=========

- Leveldown for Android, iOS. Also works on Destkop with SpiderMonkey / v8 engine
- Requires [JXcore](https://github.com/jxcore/jxcore) to run.

<img alt="LevelDB Logo" height="100" src="http://leveldb.org/img/logo.svg">

##### Installation

- To use with a mobile node application (i.e. JXcore)

```
npm install leveldown-mobile
```

- To compile for SpiderMonkey desktop or OpenWrt embedded

```
jx install leveldown-mobile  
```

A Low-level Node.js LevelDB binding
-------------------------

[![Build Status](https://secure.travis-ci.org/Level/leveldown-mobile.png)](http://travis-ci.org/Level/leveldown-mobile)

LevelDOWN was extracted from [LevelUP](https://github.com/level/levelup) and now serves as a stand-alone binding for LevelDB.

##### Remarks

- This version of Leveldown only works with JXcore and platform independent. 

- It is **strongly recommended** that you use LevelUP in preference to LevelDOWN unless you have measurable performance reasons to do so. LevelUP is optimised for usability and safety. Although we are working to improve the safety of the LevelDOWN interface it is still easy to crash your Node process if you don't do things in just the right way.

See the section on <a href="#safety">safety</a> below for details of known unsafe operations with LevelDOWN.

<a name="platforms"></a>
Tested & supported platforms
----------------------------
  * **Android**
  * **iOS**
  * **Linux** 
  * **Mac OS**
  * **Solaris** 
  * **FreeBSD**
  * **Windows**

<a name="api"></a>
## API

  * <a href="#ctor"><code><b>leveldown()</b></code></a>
  * <a href="#leveldown_open"><code><b>leveldown#open()</b></code></a>
  * <a href="#leveldown_close"><code><b>leveldown#close()</b></code></a>
  * <a href="#leveldown_put"><code><b>leveldown#put()</b></code></a>
  * <a href="#leveldown_get"><code><b>leveldown#get()</b></code></a>
  * <a href="#leveldown_del"><code><b>leveldown#del()</b></code></a>
  * <a href="#leveldown_batch"><code><b>leveldown#batch()</b></code></a>
  * <a href="#leveldown_approximateSize"><code><b>leveldown#approximateSize()</b></code></a>
  * <a href="#leveldown_getProperty"><code><b>leveldown#getProperty()</b></code></a>
  * <a href="#leveldown_iterator"><code><b>leveldown#iterator()</b></code></a>
  * <a href="#iterator_next"><code><b>iterator#next()</b></code></a>
  * <a href="#iterator_end"><code><b>iterator#end()</b></code></a>
  * <a href="#leveldown_destroy"><code><b>leveldown.destroy()</b></code></a>
  * <a href="#leveldown_repair"><code><b>leveldown.repair()</b></code></a>


--------------------------------------------------------
<a name="ctor"></a>
### leveldown(location)
<code>leveldown()</code> returns a new **LevelDOWN** instance. `location` is a String pointing to the LevelDB location to be opened.


--------------------------------------------------------
<a name="leveldown_open"></a>
### leveldown#open([options, ]callback)
<code>open()</code> is an instance method on an existing database object.

The `callback` function will be called with no arguments when the database has been successfully opened, or with a single `error` argument if the open operation failed for any reason.

#### `options`

The optional `options` argument may contain:

* `'createIfMissing'` *(boolean, default: `true`)*: If `true`, will initialise an empty database at the specified location if one doesn't already exist. If `false` and a database doesn't exist you will receive an error in your `open()` callback and your database won't open.

* `'errorIfExists'` *(boolean, default: `false`)*: If `true`, you will receive an error in your `open()` callback if the database exists at the specified location.

* `'compression'` *(boolean, default: `true`)*: If `true`, all *compressible* data will be run through the Snappy compression algorithm before being stored. Snappy is very fast and shouldn't gain much speed by disabling so leave this on unless you have good reason to turn it off.

* `'cacheSize'` *(number, default: `8 * 1024 * 1024` = 8MB)*: The size (in bytes) of the in-memory [LRU](http://en.wikipedia.org/wiki/Cache_algorithms#Least_Recently_Used) cache with frequently used uncompressed block contents. 

**Advanced options**

The following options are for advanced performance tuning. Modify them only if you can prove actual benefit for your particular application.

* `'writeBufferSize'` *(number, default: `4 * 1024 * 1024` = 4MB)*: The maximum size (in bytes) of the log (in memory and stored in the .log file on disk). Beyond this size, LevelDB will convert the log data to the first level of sorted table files. From the LevelDB documentation:

> Larger values increase performance, especially during bulk loads. Up to two write buffers may be held in memory at the same time, so you may wish to adjust this parameter to control memory usage. Also, a larger write buffer will result in a longer recovery time the next time the database is opened.

* `'blockSize'` *(number, default `4096` = 4K)*: The *approximate* size of the blocks that make up the table files. The size related to uncompressed data (hence "approximate"). Blocks are indexed in the table file and entry-lookups involve reading an entire block and parsing to discover the required entry.

* `'maxOpenFiles'` *(number, default: `1000`)*: The maximum number of files that LevelDB is allowed to have open at a time. If your data store is likely to have a large working set, you may increase this value to prevent file descriptor churn. To calculate the number of files required for your working set, divide your total data by 2MB, as each table file is a maximum of 2MB. 

* `'blockRestartInterval'` *(number, default: `16`)*: The number of entries before restarting the "delta encoding" of keys within blocks. Each "restart" point stores the full key for the entry, between restarts, the common prefix of the keys for those entries is omitted. Restarts are similar to the concept of keyframs in video encoding and are used to minimise the amount of space required to store keys. This is particularly helpful when using deep namespacing / prefixing in your keys.


--------------------------------------------------------
<a name="leveldown_close"></a>
### leveldown#close(callback)
<code>close()</code> is an instance method on an existing database object. The underlying LevelDB database will be closed and the `callback` function will be called with no arguments if the operation is successful or with a single `error` argument if the operation failed for any reason.


--------------------------------------------------------
<a name="leveldown_put"></a>
### leveldown#put(key, value[, options], callback)
<code>put()</code> is an instance method on an existing database object, used to store new entries, or overwrite existing entries in the LevelDB store.

The `key` and `value` objects may either be `String`s or Node.js `Buffer` objects. Other object types are converted to JavaScript `String`s with the `toString()` method. Keys may not be `null` or `undefined` and objects converted with `toString()` should not result in an empty-string. Values of `null`, `undefined`, `''`, `[]` and `new Buffer(0)` (and any object resulting in a `toString()` of one of these) will be stored as a zero-length character array and will therefore be retrieved as either `''` or `new Buffer(0)` depending on the type requested.

A richer set of data-types are catered for in LevelUP.

#### `options`

The only property currently available on the `options` object is `'sync'` *(boolean, default: `false`)*. If you provide a `'sync'` value of `true` in your `options` object, LevelDB will perform a synchronous write of the data; although the operation will be asynchronous as far as Node is concerned. Normally, LevelDB passes the data to the operating system for writing and returns immediately, however a synchronous write will use `fsync()` or equivalent so your callback won't be triggered until the data is actually on disk. Synchronous filesystem writes are **significantly** slower than asynchronous writes but if you want to be absolutely sure that the data is flushed then you can use `'sync': true`.

The `callback` function will be called with no arguments if the operation is successful or with a single `error` argument if the operation failed for any reason.


--------------------------------------------------------
<a name="leveldown_get"></a>
### leveldown#get(key[, options], callback)
<code>get()</code> is an instance method on an existing database object, used to fetch individual entries from the LevelDB store.

The `key` object may either be a `String` or a Node.js `Buffer` object and cannot be `undefined` or `null`. Other object types are converted to JavaScript `String`s with the `toString()` method and the resulting `String` *may not* be a zero-length. A richer set of data-types are catered for in LevelUP.

Values fetched via `get()` that are stored as zero-length character arrays (`null`, `undefined`, `''`, `[]`, `new Buffer(0)`) will return as empty-`String` (`''`) or `new Buffer(0)` when fetched with `asBuffer: true` (see below).

#### `options`

The optional `options` object may contain:

* `'fillCache'` *(boolean, default: `true`)*: LevelDB will by default fill the in-memory LRU Cache with data from a call to get. Disabling this is done by setting `fillCache` to `false`.

* `'asBuffer'` *(boolean, default: `true`)*: Used to determine whether to return the `value` of the entry as a `String` or a Node.js `Buffer` object. Note that converting from a `Buffer` to a `String` incurs a cost so if you need a `String` (and the `value` can legitimately become a UFT8 string) then you should fetch it as one with `asBuffer: true` and you'll avoid this conversion cost.

The `callback` function will be called with a single `error` if the operation failed for any reason. If successful the first argument will be `null` and the second argument will be the `value` as a `String` or `Buffer` depending on the `asBuffer` option.


--------------------------------------------------------
<a name="leveldown_del"></a>
### leveldown#del(key[, options], callback)
<code>del()</code> is an instance method on an existing database object, used to delete entries from the LevelDB store.

The `key` object may either be a `String` or a Node.js `Buffer` object and cannot be `undefined` or `null`. Other object types are converted to JavaScript `String`s with the `toString()` method and the resulting `String` *may not* be a zero-length. A richer set of data-types are catered for in LevelUP.

#### `options`

The only property currently available on the `options` object is `'sync'` *(boolean, default: `false`)*. See <a href="#leveldown_put">leveldown#put()</a> for details about this option.

The `callback` function will be called with no arguments if the operation is successful or with a single `error` argument if the operation failed for any reason.


--------------------------------------------------------
<a name="leveldown_batch"></a>
### leveldown#batch(operations[, options], callback)

CHANGELOG.md
Note that we have filtered out commits related to new tags, updating changelog and we're also not listing any merge commits, i.e. we are only listing things that changed between versions.
### 1.1.1 January 28 2016
* Bug fixes

### 1.1.0 November 12 2015

* Windows UWP support
* Chakra engine option
* Macro performance updates
### 1.0.10 July 06 2015

* leveldown-mobile native extension works also on desktop
* provide fast discovery path for devices with slow IO

### 1.0.7/1.0.9 June 29 2015

* npmignore leveldown_embedded.gyp
* fix Linux compile issue
* fix npm install
* fix Windows compile issue

### 1.0.6m June 20 2015

* Updated for Android and iOS (mostly mechanical changes)

### 1.0.5/1.0.6 May 5 2015

* [[`9064099fe7`](https://github.com/level/leveldown/commit/9064099fe7)] - pass db to abstractIterator so gc keeps it (Julian Gruber)

### 1.0.4 May 5 2015

 * [[`b550c98291`](https://github.com/level/leveldown/commit/b550c98291)] - update nan for iojs 2.0.0 (Lars-Magnus Skog)

### 1.0.3 May 2 2015

 * [[`82479b689f`](https://github.com/level/leveldown/commit/82479b689f)] - tap -> tape + faucet (Lars-Magnus Skog)
 * [[`ca9101542a`](https://github.com/level/leveldown/commit/ca9101542a)] - fix write-random.js, use leveldown instead of lmdb (Lars-Magnus Skog)
 * [[`03fbbfb99f`](https://github.com/level/leveldown/commit/03fbbfb99f)] - fix bench/db-bench.js (Lars-Magnus Skog)

### 1.0.2 Apr 26 2015
  * [[`8470a63678`](https://github.com/level/leveldown/commit/8470a63678)] - s/rvagg\/node-/level\// (Lars-Magnus Skog)
  * [[`9cbf592bea`](https://github.com/level/leveldown/commit/9cbf592bea)] - add documentation about snapshots (Max Ogden)
  * [[`b57827cd29`](https://github.com/level/leveldown/commit/b57827cd29)] - use n instead of nvm for working iojs support (Lars-Magnus Skog)
  * [[`a19927667a`](https://github.com/level/leveldown/commit/a19927667a)] - abstract-leveldown ~2.1.0 (ralphtheninja)
  * [[`95ccdf0850`](https://github.com/level/leveldown/commit/95ccdf0850)] - update logo and copyright (Lars-Magnus Skog)
  * [[`09e89d7abb`](https://github.com/level/leveldown/commit/09e89d7abb)] - updated my email (ralphtheninja)

### 1.0.1 Jan 16 2015
  * [[`6df3ecd6f5`](https://github.com/level/leveldown/commit/6df3ecd6f5)] - nan 1.5 for io.js support (Rod Vagg)
  * [[`5198231a88`](https://github.com/level/leveldown/commit/5198231a88)] - Fix LevelDB builds for modern gcc versions (Sharvil Nanavati)

### 1.0.0 Aug 26 2014
  * NAN@1.3 for Node 0.11.13+ support (@rvagg)
  * Allow writing empty values: null, undefined, '', [] and Buffer(0). Entries come out as '' or Buffer(0) (@ggreer, @juliangruber, @rvagg)
  * Fix clang build (@thlorenz)
  * Massive speed up of iterators by chunking reads (@kesla)
  * Wrap in abstract-leveldown for consistent type-checking across *DOWNs (@kesla)
  * Upgrade to LevelDB 1.17.0 (@kesla)
  * Minor memory leaks
  * Remove compile option that borked EL5 compiles
  * Switch to plain MIT license

### 0.10.2 @ Nov 30 2013

  * Apply fix by @rescrv for long-standing OSX corruption bug, https://groups.google.com/forum/#!topic/leveldb/GXhx8YvFiig (@rvagg / @rescrv)

### 0.10.1 @ Nov 21 2013

  * NAN@0.6 for Node@0.11.6 support, v8::Local<T>::New(val) rewritten to
    NanNewLocal<T>(val) (@rvagg)

### 0.10.0 @ Nov 18 2013

  * Fix array-batch memory leak, levelup/#171 (@rvagg)
  * Fix chained-batch write() segfaults, details in #73, (@rvagg and
    @mcollina)
  * Remove `Persistent` references for all `batch()` operations as
    `WriteBatch` takes an explicit copy of the data (@mcollina and
@rvagg)
  * Upgrade to Snappy 1.1.1 (@rvagg and @no9)
  * Upgrade to NAN@0.5.x (@rvagg)
  * Switch all `callback->Call()`s to `node::MakeCallback()` to properly
    support Node.js domains (@rvagg)
  * Properly enable compression by default (@Kyotoweb)
  * Enable LevelDB's BloomFilter (@Kyotoweb)
  * Upgrade to AbstractLevelDOWN@0.11.x for testing (@rvagg)
  * Add new simple batch() leak tester (@rvagg)

### 0.9.2 @ Nov 02 2013

  * Minor fixes to support Node 0.11.8 and new Linux gcc (warnings) (@rvagg)

### 0.9.1 @ Oct 03 2013

  * Include port_uv.h for Windows compile, added test to suite to make sure this happens every time LevelDB is upgraded (@rvagg)

### 0.9.0 @ Oct 01 2013

  * Upgrade from LevelDB@0.11.0 to LevelDB@0.14.0, includes change from .sst to .ldb file extension for SST files (@rvagg)

### 0.8.3 @ Sept 18 2013

  * Upgrade to nan@0.4.0, better support for latest Node master & support for installing within directory structures containing spaces in directory names (@rvagg)

### 0.8.2 @ Sept 2 2013

  * FreeBSD support (@rvagg, @kelexel)

### 0.8.1 @ Sept 1 2013

  * Fixed some minor V8-level leaks (@rvagg)

### 0.8.0 @ Aug 19 2013

  * Added `gt`, `lt`, `gte`, `lte` for iterators (@dominictarr)
  * Switch to NAN as an npm dependency (@rvagg)

### 0.7.0 @ Aug 11 2013

  * Added @pgte to contributors list
  * (very) Minor perf improvements in C++ (@mscdex)
  * Use NAN <https://github.com/rvagg/nan> for Node 0.8->0.11 compatibility

### 0.6.2 @ Jul 07 2013

  * Compatibility for Node 0.11.3, breaks compatibility with 0.11.2

### 0.6.1 @ Jun 15 2013

  * Fix broken Windows compile, apply port-uv patch to LevelDB's port.h (@rvagg)

### 0.6.0 @ Jun 14 2013

  * Upgrade to LevelDB 1.11.0, some important bugfixes: https://groups.google.com/forum/#!topic/leveldb/vS1JvmGlp4E

### 0.5.0 @ May 18 2013

  * Bumped major version for db.getProperty() addition (should have been done in 0.4.4) (@rvagg)
  * Disallow batch() operations after a write() (@rvagg)

### 0.4.4 @ May 18 2013


src/database.cc
#if defined(JS_ENGINE_V8) or defined(JS_ENGINE_MOZJS) or \
    defined(JS_ENGINE_CHAKRA)
/* Copyright (c) 2012-2015 LevelDOWN contributors
 * See list at <https://github.com/level/leveldown#contributing>
 * MIT License <https://github.com/level/leveldown/blob/master/LICENSE.md>
 */

#include <node.h>
#include <node_buffer.h>

#include <leveldb/db.h>
#include <leveldb/write_batch.h>

#include "leveldown.h"
#include "database.h"
#include "async.h"
#include "database_async.h"
#include "batch.h"
#include "iterator.h"
#include "jx_persistent_store.h"

namespace leveldown {

jxcore::ThreadStore<JS_PERSISTENT_FUNCTION_TEMPLATE> Database::jx_persistent;

Database::Database(const JS_HANDLE_VALUE& from)
    : db(NULL),
      currentIteratorId(0),
      pendingCloseWorker(NULL),
      blockCache(NULL),
      filterPolicy(NULL) {
  location.SetFromHandle(from);
};

Database::~Database() {
  if (db != NULL) delete db;
  location.Dispose();
};

/* Calls from worker threads, NO V8 HERE *****************************/

leveldb::Status Database::OpenDatabase(leveldb::Options* options) {
  return leveldb::DB::Open(*options, *location, &db);
}

leveldb::Status Database::PutToDatabase(leveldb::WriteOptions* options,
                                        leveldb::Slice key,
                                        leveldb::Slice value) {
  return db->Put(*options, key, value);
}

leveldb::Status Database::GetFromDatabase(leveldb::ReadOptions* options,
                                          leveldb::Slice key,
                                          std::string& value) {
  return db->Get(*options, key, &value);
}

leveldb::Status Database::DeleteFromDatabase(leveldb::WriteOptions* options,
                                             leveldb::Slice key) {
  return db->Delete(*options, key);
}

leveldb::Status Database::WriteBatchToDatabase(leveldb::WriteOptions* options,
                                               leveldb::WriteBatch* batch) {
  return db->Write(*options, batch);
}

uint64_t Database::ApproximateSizeFromDatabase(const leveldb::Range* range) {
  uint64_t size;
  db->GetApproximateSizes(range, 1, &size);
  return size;
}

void Database::GetPropertyFromDatabase(const leveldb::Slice& property,
                                       std::string* value) {

  db->GetProperty(property, value);
}

leveldb::Iterator* Database::NewIterator(leveldb::ReadOptions* options) {
  return db->NewIterator(*options);
}

const leveldb::Snapshot* Database::NewSnapshot() { return db->GetSnapshot(); }

void Database::ReleaseSnapshot(const leveldb::Snapshot* snapshot) {
  return db->ReleaseSnapshot(snapshot);
}

void Database::ReleaseIterator(uint32_t id) {
  // called each time an Iterator is End()ed, in the main thread
  // we have to remove our reference to it and if it's the last iterator
  // we have to invoke a pending CloseWorker if there is one
  // if there is a pending CloseWorker it means that we're waiting for
  // iterators to end before we can close them
  iterators.erase(id);
  if (iterators.empty() && pendingCloseWorker != NULL) {
    NanAsyncQueueWorker((AsyncWorker*)pendingCloseWorker);
    pendingCloseWorker = NULL;
  }
}

void Database::CloseDatabase() {
  delete db;
  db = NULL;
  if (blockCache) {
    delete blockCache;
    blockCache = NULL;
  }
  if (filterPolicy) {
    delete filterPolicy;
    filterPolicy = NULL;
  }
}

/* V8 exposed functions *****************************/

JS_METHOD(Database, New) {
  JS_CLASS_NEW_INSTANCE(obj, Database);
  JS_LOCAL_OBJECT arg0 = JS_VALUE_TO_OBJECT(args.GetItem(0));
  Database* db = new Database(arg0);
  db->Wrap(obj);

  RETURN_PARAM(obj);
}
JS_METHOD_END

JS_HANDLE_VALUE Database::NewInstance(JS_LOCAL_STRING& location) {
  JS_ENTER_SCOPE_COM();
  JS_DEFINE_STATE_MARKER(com);

  JS_LOCAL_OBJECT instance;

  JS_LOCAL_FUNCTION_TEMPLATE constructorHandle =
      JS_TYPE_TO_LOCAL_FUNCTION_TEMPLATE(
          Database::jx_persistent.templates[com->threadId]);

  JS_HANDLE_VALUE argv[] = {location};

  JS_LOCAL_FUNCTION cons = JS_GET_FUNCTION(constructorHandle);
  instance = JS_NEW_INSTANCE(cons, 1, argv);

  return JS_LEAVE_SCOPE(instance);
}

JS_METHOD(Database, Open) {
  LD_METHOD_SETUP_COMMON(open, 0, 1)

  JS_LOCAL_VALUE obj;

  bool createIfMissing = true;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("createIfMissing"))) {
    obj = JS_GET_NAME(optionsObj, JS_STRING_ID("createIfMissing"));
    createIfMissing = BOOLEAN_TO_STD(obj);
  }

  bool errorIfExists = false;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("errorIfExists"))) {
    obj = JS_GET_NAME(optionsObj, JS_STRING_ID("errorIfExists"));
    errorIfExists = BOOLEAN_TO_STD(obj);
  }

  bool compression = true;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("compression"))) {
    obj = JS_GET_NAME(optionsObj, JS_STRING_ID("compression"));
    compression = BOOLEAN_TO_STD(obj);
  }

  uint32_t cacheSize = 8 << 20;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("cacheSize"))) {
    obj = JS_GET_NAME(optionsObj, JS_STRING_ID("cacheSize"));
    cacheSize = NUMBER_TO_STD(obj);
  }

  uint32_t writeBufferSize = 4 << 20;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("writeBufferSize"))) {
    obj = JS_GET_NAME(optionsObj, JS_STRING_ID("writeBufferSize"));
    writeBufferSize = NUMBER_TO_STD(obj);
  }

  uint32_t blockSize = 4096;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("blockSize"))) {
    obj = JS_GET_NAME(optionsObj, JS_STRING_ID("blockSize"));
    blockSize = NUMBER_TO_STD(obj);
  }

  uint32_t maxOpenFiles = 1000;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("maxOpenFiles"))) {
    obj = JS_GET_NAME(optionsObj, JS_STRING_ID("maxOpenFiles"));
    maxOpenFiles = NUMBER_TO_STD(obj);
  }

  uint32_t blockRestartInterval = 16;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("blockRestartInterval"))) {
    obj = JS_GET_NAME(optionsObj, JS_STRING_ID("blockRestartInterval"));
    blockRestartInterval = NUMBER_TO_STD(obj);
  }

  database->blockCache = leveldb::NewLRUCache(cacheSize);
  database->filterPolicy = leveldb::NewBloomFilterPolicy(10);

  OpenWorker* worker = new OpenWorker(
      database, new NanCallback(callback), database->blockCache,
      database->filterPolicy, createIfMissing, errorIfExists, compression,
      writeBufferSize, blockSize, maxOpenFiles, blockRestartInterval);
  // persist to prevent accidental GC
  JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
  worker->SaveToPersistent("database", _this);
  NanAsyncQueueWorker(worker);
}
JS_METHOD_END

JS_METHOD(Database, Close) {
  LD_METHOD_SETUP_COMMON_ONEARG(close)

  CloseWorker* worker = new CloseWorker(database, new NanCallback(callback));
  // persist to prevent accidental GC
  JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
  worker->SaveToPersistent("database", _this);

  if (!database->iterators.empty()) {
    // yikes, we still have iterators open! naughty naughty.
    // we have to queue up a CloseWorker and manually close each of them.
    // the CloseWorker will be invoked once they are all cleaned up
    database->pendingCloseWorker = worker;

    for (std::map<uint32_t, leveldown::Iterator*>::iterator it =
             database->iterators.begin();
         it != database->iterators.end(); ++it) {

      // for each iterator still open, first check if it's already in
      // the process of ending (ended==true means an async End() is
      // in progress), if not, then we call End() with an empty callback
      // function and wait for it to hit ReleaseIterator() where our
      // CloseWorker will be invoked

      leveldown::Iterator* iterator = it->second;

      if (!iterator->ended) {
        JS_LOCAL_OBJECT obj = NanObjectWrapHandle(iterator);
        JS_LOCAL_VALUE obj_end;
        if (JS_HAS_NAME(obj, JS_STRING_ID("end"))) {
          obj_end = JS_GET_NAME(obj, JS_STRING_ID("end"));
        }

        if (JS_IS_FUNCTION(obj_end)) {
          JS_LOCAL_FUNCTION end = JS_TYPE_AS_FUNCTION(obj_end);
          JS_LOCAL_FUNCTION_TEMPLATE tmp = JS_NEW_EMPTY_FUNCTION_TEMPLATE();

          JS_LOCAL_VALUE argv[] = {JS_GET_FUNCTION(tmp)};
          NanMakeCallback(obj, end, 1, argv);
        } else {
          THROW_EXCEPTION(
              "database instance doesn't have end function defined");
        }
      }
    }
  } else {
    NanAsyncQueueWorker(worker);
  }
}
JS_METHOD_END

JS_METHOD(Database, Put) {
  LD_METHOD_SETUP_COMMON(put, 2, 3)

  JS_LOCAL_OBJECT keyHandle;
  if (!args.IsNull(0) && !args.IsUndefined(0))
    keyHandle = JS_VALUE_TO_OBJECT(GET_ARG(0));

  JS_LOCAL_OBJECT valueHandle;
  
  if (!args.IsNull(1) && !args.IsUndefined(1))  
    valueHandle = JS_VALUE_TO_OBJECT(GET_ARG(1));
  
  LD_STRING_OR_BUFFER_TO_SLICE(key, keyHandle, key)
  LD_STRING_OR_BUFFER_TO_SLICE(value, valueHandle, value)

  bool sync = false;
  if (!JS_IS_EMPTY(optionsObj) && JS_HAS_NAME(optionsObj, JS_STRING_ID("sync"))) {
    JS_LOCAL_VALUE obj_sync = JS_GET_NAME(optionsObj, JS_STRING_ID("sync"));
    sync = BOOLEAN_TO_STD(obj_sync);
  }

  WriteWorker* worker =
      new WriteWorker(database, new NanCallback(callback), key, value, sync,
                      keyHandle, valueHandle);

  // persist to prevent accidental GC
  JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
  worker->SaveToPersistent("database", _this);
  NanAsyncQueueWorker(worker);
}
JS_METHOD_END

JS_METHOD(Database, Get) {
  LD_METHOD_SETUP_COMMON(get, 1, 2)

  JS_LOCAL_OBJECT keyHandle = JS_VALUE_TO_OBJECT(args.GetItem(0));
  LD_STRING_OR_BUFFER_TO_SLICE(key, keyHandle, key)

  bool asBuffer = true;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("asBuffer"))) {
    JS_LOCAL_VALUE obj_sync = JS_GET_NAME(optionsObj, JS_STRING_ID("asBuffer"));
    asBuffer = BOOLEAN_TO_STD(obj_sync);
  }

  bool fillCache = true;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("fillCache"))) {
    JS_LOCAL_VALUE obj_sync =
        JS_GET_NAME(optionsObj, JS_STRING_ID("fillCache"));
    fillCache = BOOLEAN_TO_STD(obj_sync);
  }

  ReadWorker* worker = new ReadWorker(database, new NanCallback(callback), key,
                                      asBuffer, fillCache, keyHandle);
  // persist to prevent accidental GC
  JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
  worker->SaveToPersistent("database", _this);
  NanAsyncQueueWorker(worker);
}
JS_METHOD_END

JS_METHOD(Database, Delete) {
  LD_METHOD_SETUP_COMMON(del, 1, 2)

  JS_LOCAL_OBJECT keyHandle = JS_VALUE_TO_OBJECT(args.GetItem(0));
  LD_STRING_OR_BUFFER_TO_SLICE(key, keyHandle, key)

  bool sync = false;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("sync"))) {
    JS_LOCAL_VALUE obj_sync = JS_GET_NAME(optionsObj, JS_STRING_ID("sync"));
    sync = BOOLEAN_TO_STD(obj_sync);
  }

  DeleteWorker* worker = new DeleteWorker(database, new NanCallback(callback),
                                          key, sync, keyHandle);
  // persist to prevent accidental GC
  JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
  worker->SaveToPersistent("database", _this);
  NanAsyncQueueWorker(worker);
}
JS_METHOD_END

JS_METHOD(Database, Batch) {
  if ((args.Length() == 0 || args.Length() == 1) && !args.IsArray(0)) {
    JS_LOCAL_OBJECT optionsObj;
    if (args.Length() > 0 && args.IsObject(0)) {
      optionsObj = JS_VALUE_TO_OBJECT(args.GetItem(0));
    }

    JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
    RETURN_PARAM(Batch::NewInstance(_this, optionsObj));
  }

  LD_METHOD_SETUP_COMMON(batch, 1, 2)

  bool sync = false;
  if (JS_HAS_NAME(optionsObj, JS_STRING_ID("sync"))) {
    JS_LOCAL_VALUE obj_sync = JS_GET_NAME(optionsObj, JS_STRING_ID("sync"));
    sync = BOOLEAN_TO_STD(obj_sync);
  }

  JS_HANDLE_ARRAY array = JS_TYPE_AS_ARRAY(args.GetItem(0));

  leveldb::WriteBatch* batch = new leveldb::WriteBatch();
  bool hasData = false;

  unsigned ln = JS_GET_ARRAY_LENGTH(array);
  for (unsigned int i = 0; i < ln; i++) {
    JS_LOCAL_VALUE val = JS_GET_INDEX(array, i);
    if (!JS_IS_OBJECT(val)) continue;
    JS_LOCAL_OBJECT obj = JS_VALUE_TO_OBJECT(val);

    JS_LOCAL_VALUE keyBuffer = JS_GET_NAME(obj, JS_STRING_ID("key"));
    JS_LOCAL_VALUE type = JS_GET_NAME(obj, JS_STRING_ID("type"));

    jxcore::JXString str_type(type);
    if (strcmp(*str_type, "del") == 0) {
      LD_STRING_OR_BUFFER_TO_SLICE(key, keyBuffer, key)

      batch->Delete(key);
      if (!hasData) hasData = true;

      DisposeStringOrBufferFromSlice(keyBuffer, key);
    } else if (strcmp(*str_type, "put") == 0) {
      JS_LOCAL_VALUE valueBuffer = JS_GET_NAME(obj, STD_TO_STRING("value"));

      LD_STRING_OR_BUFFER_TO_SLICE(key, keyBuffer, key)
      LD_STRING_OR_BUFFER_TO_SLICE(value, valueBuffer, value)
      batch->Put(key, value);
      if (!hasData) hasData = true;

      DisposeStringOrBufferFromSlice(keyBuffer, key);
      DisposeStringOrBufferFromSlice(valueBuffer, value);
    }
  }

  // don't allow an empty batch through
  if (hasData) {
    BatchWorker* worker =
        new BatchWorker(database, new NanCallback(callback), batch, sync);
    // persist to prevent accidental GC
    JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
    worker->SaveToPersistent("database", _this);
    NanAsyncQueueWorker(worker);
  } else {
    LD_RUN_CALLBACK(callback, 0, NULL);
  }
}
JS_METHOD_END

JS_METHOD(Database, ApproximateSize) {
  JS_LOCAL_OBJECT startHandle = JS_VALUE_TO_OBJECT(args.GetItem(0));
  JS_LOCAL_OBJECT endHandle = JS_VALUE_TO_OBJECT(args.GetItem(1));

  LD_METHOD_SETUP_COMMON(approximateSize, -1, 2)

  LD_STRING_OR_BUFFER_TO_SLICE(start, startHandle, start)
  LD_STRING_OR_BUFFER_TO_SLICE(end, endHandle, end)

  ApproximateSizeWorker* worker = new ApproximateSizeWorker(
      database, new NanCallback(callback), start, end, startHandle, endHandle);
  // persist to prevent accidental GC
  JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
  worker->SaveToPersistent("database", _this);
  NanAsyncQueueWorker(worker);
}
JS_METHOD_END

JS_METHOD(Database, GetProperty) {
  if (args.Length() < 1) {
    THROW_EXCEPTION("Database::GetProperty expects a parameter");
  }

  JS_LOCAL_VALUE propertyHandle = JS_VALUE_TO_OBJECT(args.GetItem(0));
  JS_LOCAL_FUNCTION callback;

  LD_STRING_OR_BUFFER_TO_SLICE(property, propertyHandle, property)

  leveldown::Database* database =
      node::ObjectWrap::Unwrap<leveldown::Database>(args.This());

  std::string* value = new std::string();
  database->GetPropertyFromDatabase(property, value);
  JS_LOCAL_STRING returnValue =
      UTF8_TO_STRING_WITH_LENGTH(value->c_str(), value->length());
  delete value;
  delete[] property.data();

  RETURN_PARAM(returnValue);
}
JS_METHOD_END

JS_METHOD(Database, Iterator) {
  Database* database = node::ObjectWrap::Unwrap<Database>(args.This());

  JS_LOCAL_OBJECT optionsObj;
  if (args.Length() > 0 && args.IsObject(0)) {
    optionsObj = JS_VALUE_TO_OBJECT(args.GetItem(0));
  }

  // each iterator gets a unique id for this Database, so we can
  // easily store & lookup on our `iterators` map
  uint32_t id = database->currentIteratorId++;
  JS_TRY_CATCH(try_catch);

  JS_LOCAL_OBJECT _this = JS_TYPE_TO_LOCAL_OBJECT(args.This());
  JS_LOCAL_OBJECT iteratorHandle =
      Iterator::NewInstance(_this, STD_TO_NUMBER(id), optionsObj);
  if (try_catch.HasCaught()) {
    THROW_EXCEPTION("Fatal Error in Database::Iterator!");
  }

  leveldown::Iterator* iterator =
      node::ObjectWrap::Unwrap<leveldown::Iterator>(iteratorHandle);

  database->iterators[id] = iterator;

  RETURN_PARAM(iteratorHandle);
}
JS_METHOD_END

}  // namespace leveldown
#endif
